Common/src/linear_algebra/CSysVector.cpp

```diff
@@ -26,472 +26,257 @@
 
-template<class ScalarType>
-CSysVector<ScalarType> & CSysVector<ScalarType>::operator/=(const ScalarType & val) {
+  PARALLEL_FOR
+  for(auto i=0ul; i<nElm; i++) vec_val[i] *= val;
 
-  for (unsigned long i = 0; i < nElm; i++)
-    vec_val[i] /= val;
   return *this;
 }
 
 template<class ScalarType>
-ScalarType CSysVector<ScalarType>::norm() const {
-
-  /*--- just call dotProd on this*, then sqrt ---*/
-  ScalarType val = dotProd(*this, *this);
-  if (val < 0.0) {
-    SU2_MPI::Error("Inner product of CSysVector is negative", CURRENT_FUNCTION);
-  }
-  return sqrt(val);
-}
-
-template<class ScalarType>
-void CSysVector<ScalarType>::CopyToArray(ScalarType* u_array) {
-
-  for (unsigned long i = 0; i < nElm; i++)
-    u_array[i] = vec_val[i];
-}
-
-template<class ScalarType>
-void CSysVector<ScalarType>::AddBlock(unsigned long val_ipoint, ScalarType *val_residual) {
-  unsigned short iVar;
-
-  for (iVar = 0; iVar < nVar; iVar++)
-    vec_val[val_ipoint*nVar+iVar] += val_residual[iVar];
-}
-
-template<class ScalarType>
-void CSysVector<ScalarType>::SubtractBlock(unsigned long val_ipoint, ScalarType *val_residual) {
-  unsigned short iVar;
-
-  for (iVar = 0; iVar < nVar; iVar++)
-    vec_val[val_ipoint*nVar+iVar] -= val_residual[iVar];
-}
-
-template<class ScalarType>
-void CSysVector<ScalarType>::SetBlock(unsigned long val_ipoint, ScalarType *val_residual) {
-  unsigned short iVar;
+CSysVector<ScalarType> & CSysVector<ScalarType>::operator/=(ScalarType val) {
 
-  for (iVar = 0; iVar < nVar; iVar++)
-    vec_val[val_ipoint*nVar+iVar] = val_residual[iVar];
-}
-
-template<class ScalarType>
-void CSysVector<ScalarType>::SetBlock(unsigned long val_ipoint, unsigned short val_var, ScalarType val_residual) {
-
-  vec_val[val_ipoint*nVar+val_var] = val_residual;
-}
+  PARALLEL_FOR
+  for(auto i=0ul; i<nElm; i++) vec_val[i] /= val;
 
-template<class ScalarType>
-void CSysVector<ScalarType>::SetBlock_Zero(unsigned long val_ipoint) {
-  unsigned short iVar;
-
-  for (iVar = 0; iVar < nVar; iVar++)
-    vec_val[val_ipoint*nVar+iVar] = 0.0;
-}
-
-template<class ScalarType>
-void CSysVector<ScalarType>::SetBlock_Zero(unsigned long val_ipoint, unsigned short val_var) {
-    vec_val[val_ipoint*nVar+val_var] = 0.0;
+  return *this;
 }
 
 template<class ScalarType>
-ScalarType CSysVector<ScalarType>::GetBlock(unsigned long val_ipoint, unsigned short val_var) {
-  return vec_val[val_ipoint*nVar + val_var];
-}
+void CSysVector<ScalarType>::CopyToArray(ScalarType* u_array) const {
 
-template<class ScalarType>
-ScalarType *CSysVector<ScalarType>::GetBlock(unsigned long val_ipoint) {
-  return &vec_val[val_ipoint*nVar];
+  PARALLEL_FOR
+  for(auto i=0ul; i<nElm; i++) u_array[i] = vec_val[i];
 }
 
 template<class ScalarType>
-template<class T>
-void CSysVector<ScalarType>::PassiveCopy(const CSysVector<T>& other) {
+ScalarType CSysVector<ScalarType>::dot(const CSysVector<ScalarType> & u) const {
+#if !defined(CODI_FORWARD_TYPE) && !defined(CODI_REVERSE_TYPE)
```

Comment:
Readability wise, the dot product operation (now a member of CSysVector) is as bad as it can get as we need to perform a reduction over threads and ranks.